src/compiler/checker.ts

```diff
@@ -1042,12 +1042,52 @@ namespace ts {
         let _jsxNamespace: __String;
         let _jsxFactoryEntity: EntityName | undefined;
 
-        const subtypeRelation = new Map<string, RelationComparisonResult>();
-        const strictSubtypeRelation = new Map<string, RelationComparisonResult>();
-        const assignableRelation = new Map<string, RelationComparisonResult>();
-        const comparableRelation = new Map<string, RelationComparisonResult>();
-        const identityRelation = new Map<string, RelationComparisonResult>();
-        const enumRelation = new Map<string, RelationComparisonResult>();
+        class InfiniMap<K, V> {
+            private next?: InfiniMap<K, V>;
+            private inner: ESMap<K, V>;
+            constructor() {
+                this.inner = new Map();
+            }
+            get(key: K): V | undefined {
+                return this.inner.has(key) ? this.inner.get(key) : this.next?.get(key);
+            }
+            set(key: K, value: V): this {
+                if (this.inner.size > ((2 ** 24) - 1)) {
+                    this.next ||= new InfiniMap();
+                    this.next.set(key, value);

```

Comment:
either `Debug.assert(!this.inner.has(key))` or set accordingly in inner map or next map ?

Reply:
The assertion is probably technically better, if not *technically* emulating normal map behavior, since our relationship caches are treated as set-once read-many.

Reply:
Turns out the cache is mutable - intentionally - to handle storing error report state for the relation. Don't really like it (nested recursive calls to `checkTypeRelatedTo` make me question the correctness of it), but so it is, not gunna change that now. So looks like it gets to handle the key presence rather than asserting it.